### PR TITLE
Fix #2578: (part 2) Make for-generators filter only if prefixed with `case`.

### LINKED ISSUE #2578: Scala Wart: Conflating total destructuring with partial pattern-matching
Opening this issue, as suggested by Martin, to provide a place to discuss the individual warts brought up in the blog post [Warts of the Scala Programming Language](http://www.lihaoyi.com/post/WartsoftheScalaProgrammingLanguage.html) and the possibility of mitigating/fixing them in Dotty (and perhaps later in Scala 2.x). These are based on Scala 2.x behavior, which I understand Dotty follows closely, apologies in advance if it has already been fixed

--------------------------------------------------------------------------------

The following example, also from 
[Lincon Atkinson's blog](http://latkin.org/blog/2017/05/02/when-the-scala-compiler-doesnt-help/), 
compiles without warning and fails with an exception at runtime:

```scala
@ val (a, b, c) = if (true) "bar" else Some(10)
scala.MatchError: bar (of class java.lang.String)
  $sess.cmd105$.<init>(cmd105.sc:1)
  $sess.cmd105$.<clinit>(cmd105.sc:-1)
```

The basic problem here is that when Scala sees `val (a, b, c) = ...`, it doesn't
mean which of two things you mean:

1. Help me extract the values from `...`, and help me check that it's a tuple
2. Help me extract the values from `...`, and fail at runtime if it is not a tuple

Currently, it assumes the latter, in all cases. 

That makes any sort of "destructuring assignment" unchecked, and thus extremely 
unsafe.

The above example at least happily fails with an exception, but the following 
exhibits the same problem, but instead truncates your data silently, losing the
`5`:
 
```scala
@ for((a, b) <- Seq(1 -> 2, 3 -> 4, 5)) yield a + " " +  b
res107: Seq[String] = List("1 2", "3 4")
```

Though the following also fails with an exception:
```scala
@ Seq(1 -> 2, 3 -> 4, 5).map{case (a, b) => a + " " + b}
scala.MatchError: 5 (of class java.lang.Integer)
  $sess.cmd108$.$anonfun$res108$1(cmd108.sc:1)
  scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
  scala.collection.immutable.List.foreach(List.scala:389)
  scala.collection.TraversableLike.map(TraversableLike.scala:234)
  scala.collection.TraversableLike.map$(TraversableLike.scala:227)
  scala.collection.immutable.List.map(List.scala:295)
  $sess.cmd108$.<init>(cmd108.sc:1)
  $sess.cmd108$.<clinit>(cmd108.sc:-1)
```

While interpretation #2 makes sense in `match` blocks and partial-functions,
where you expect to "fall through" to the next handler if it doesn't match, it
doesn't make much sense in cases like this where there *is nowhere to fall
through to*.

The correct solution would look something like this:

- By default, assume the user wants 1. "Help me extract the values from `...`, 
  and help me check that it's a tuple"

- Require a special keyword if the user wants 2. "Help me extract the values 
  from `...`, and fail at runtime if it is not a tuple"

A possible syntax might be using `case`, which Scala developers already 
associate with partial functions and pattern matches: 

```scala
for(case (a, b) <- Seq(1 -> 2, 3 -> 4, 5)) yield a + " " +  b

case val (a, b, c) = if (true) "bar" else Some(10)
```

This would indicate that you want to perform an "partial fail at runtime" 
match, and the earlier non-`case` examples:

```scala
for((a, b) <- Seq(1 -> 2, 3 -> 4, 5)) yield a + " " +  b

val (a, b, c) = if (true) "bar" else Some(10)
```

Could then verify that the pattern match is complete, otherwise fail at
compile time.

PostScript:

I noticed [here](https://www.scala-lang.org/files/archive/spec/2.11/08-pattern-matching.html#irrefutable-patterns) that the Scala language spec already has words in it that talk about irrefutable patterns, which seem to match what I want these for-comprehension and val-destructuring cases to require. Whether those words mean anything, I do not actually know

Yawar Amin has noted that in the Rust language, the cases which closely mirror those in Scala (`let` destructuring, and `while`-loop/`if` destructuring) do have a requirement of the destructuring patterns being irrefutable https://doc.rust-lang.org/beta/book/second-edition/ch18-02-refutability.html

Here's a 2.12 failure mode in for-comprehensions that's nonsensical, and is fundamentally caused by this issue:

```scala
lihaoyi ~$ amm
Loading...
Welcome to the Ammonite Repl 0.9.7
(Scala 2.12.2 Java 1.8.0_112)
If you like Ammonite, please support our development at www.patreon.com/lihaoyi
@ for((a, b) <- Right((1, 2))) yield a + b
cmd0.sc:1: value withFilter is not a member of scala.util.Right[Nothing,(Int, Int)]
val res0 = for((a, b) <- Right((1, 2))) yield a + b
                              ^
cmd0.sc:1: type mismatch;
 found   : Any
 required: String
val res0 = for((a, b) <- Right((1, 2))) yield a + b
                                                  ^
Compilation Failed
```

### PR DESCRIPTION
**Pull request by odersky**
This is a complement to #6389. Together, they address all aspects of #2578. 

The checking part of this is currently only enabled under -strict to allow cross-compilation with Scala 2. As things currently stand, this means it will be made standard in Scala 3.1.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Checking.scala [hunk 1]
```diff
@@ -594,6 +597,55 @@ trait Checking {
...
+
+    def check(pat: Tree, pt: Type): Boolean = {
+      if (pt <:< pat.tpe)
+        true
+      else {
+        var reportedPt = pt.dropAnnot(defn.UncheckedAnnot)
+        if (!pat.tpe.isSingleton) reportedPt = reportedPt.widen
+        val problem = if (pat.tpe <:< reportedPt) "is more specialized than" else "does not match"
+        val fix = if (isPatDef) "`: @unchecked` after" else "`case ` before"
+        ctx.errorOrMigrationWarning(
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/typer/Checking.scala]: I think we should either make this a warning, or turn the corresponding pattern matching warning into type error.

Comment (AleksanderBG) [reply compiler/src/dotty/tools/dotc/typer/Checking.scala]: I'm actually all for making inexhaustive match an error and requiring an `@inexhaustive` annotation to accept it.

File: docs/docs/reference/changed-features/pattern-bindings.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+In Scala 2, pattern bindings in `val` definitions and `for` expressions are
+loosely typed. Potentially failing matches are still accepted at compile-time,
+but may influence the program's runtime behavior.
+From Scala 3.1 on, type checking rules will be tightened so that errors are reported at compile-time instead.
+
+## Bindings in Pattern Definitions
+
+```scala
+  val xs: List[Any] = List(1, 2, 3)
+  val (x: String) :: _ xs     // error: pattern's type String is more specialized
```
Comment (japgolly) [inline docs/docs/reference/changed-features/pattern-bindings.md]: missing an `=` here

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 3]
```diff
@@ -1341,41 +1363,47 @@ object desugar {
         }
       }
 
-      def isIrrefutableGenFrom(gen: GenFrom): Boolean =
-        gen.isInstanceOf[IrrefutableGenFrom] ||
-        IdPattern.unapply(gen.pat).isDefined ||
-        isIrrefutable(gen.pat, gen.expr)
+      def needsNoFilter(gen: GenFrom): Boolean =
+        if (gen.checkMode == GenCheckMode.FilterAlways) // pattern was prefixed by `case`
+          isIrrefutable(gen.pat, gen.expr)
```
Comment (AleksanderBG) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: As a matter of principle, I think that if the user explicitly requested a "filtering binding" by writing `case`, then it is less surprising if we always emit a `withFilter` call. With `zipped`, we've already seen that emitting or not emitting the call can have unobvious consequences on typechecking. I think it'd be for the best if we just let the user decide whether the call is emitted or not.

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala [hunk 4]
```diff
@@ -925,6 +936,22 @@ object desugar {
     }
   }
 
+  /** The selector of a match, which depends of the given `checkMode`.
```
Comment (AleksanderBG) [inline compiler/src/dotty/tools/dotc/ast/Desugar.scala]: The fact that `@unchecked` is also added to irrefutable patterns really confused me until I understood that annotating the selector only affects exhaustivity checks and not irrefutability checks. Perhaps we could note this in the doc?

```suggestion
  /** The selector of a match, which depends of the given `checkMode`.
   *
   * The @unchecked annotation is added whenever `checkMode` is not `Exhaustive` to silence
   * unnecessary inexhaustive match warnings.
```
```

File: compiler/src/dotty/tools/dotc/ast/Desugar.scala (hunk 3 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/ast/Desugar.scala]: Good point. I'll change it.

**Pull request merged**